cpu/cortexm_common/thread_arch.c

```diff
@@ -366,3 +367,75 @@ void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
     /* {r0-r3,r12,LR,PC,xPSR,s0-s15,FPSCR} are restored automatically on exception return */
     );
 }
+
+#ifdef MODULE_CORTEXM_SVC
+void svc_dispatch(unsigned int *svc_args);
+
+void __attribute__((naked)) __attribute__((used)) isr_svc(void)
+{
+    /* these two variants do exactly the same, but Cortex-M3 can use Thumb2
+     * conditional execution, which are a bit faster. */
+
+    /* TODO: currently, cpu_switch_context_exit() is used to start threading
+     * from kernel_init(), which executes on MSP.  That could probably be
+     * rewritten to not use the supervisor call at all. Then we can assume that
+     * svc is only used by threads, saving a couple of instructions. /Kaspar
+     */
+
+#if defined(CPU_ARCH_CORTEX_M0) || defined(CPU_ARCH_CORTEX_M0PLUS) \
+    || defined(CPU_ARCH_CORTEX_M23)
+    __asm__ volatile (
+    ".thumb_func            \n"
+    "movs   r0, #4          \n" /* if bit4(lr) == 1):       */
+    "mov    r1, lr          \n"
+    "tst    r0, r1          \n"
+    "beq    came_from_msp   \n" /*     goto came_from_msp   */
+    "mrs    r0, psp         \n" /* r0 = psp                 */
+    "b      svc_dispatch    \n" /* return svc_dispatch(r0)  */
+    "came_from_msp:         \n"
+    "mrs    r0, msp         \n" /* r0 = msp                 */
+    "b      svc_dispatch    \n" /* return svc_dispatch(r0)  */
+    );
+#else
+    __asm__ volatile (
+    ".thumb_func            \n"
+    "tst    lr, #4          \n" /* switch bit4(lr) == 1):   */
+    "ite    eq              \n"
+    "mrseq  r0, msp         \n" /* case 1: r0 = msp         */
+    "mrsne  r0, psp         \n" /* case 0: r0 = psp         */
+    "b      svc_dispatch    \n" /* return svc_dispatch()    */
+    );
+#endif
+}
+
+void svc_dispatch(unsigned int *svc_args)
```

Comment:
Just nitpicking here, buy why `dispatch` and not `svc_handler`?

Reply:
Well, there are three "classes" involved:

1. the actual exception handler (`isr_svc()`)
2. something that determines what function to call (does dispatching depending on svc number, `svc_dispatch()`)
3. handlers for some svc numbers (we don't have those yet).

That's why it is called "dispatch". ;) makes sense?

Reply:
Yep!